sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/BaseAuthorizationTokenProvider.java

```diff
@@ -203,25 +202,28 @@ private Mac getMacInstance() {
 
         //  Master key has changed, or this is the first time we are getting mac instance
         if (masterKeyLatestHashCode != this.masterKeyHashCode) {
-            byte[] masterKeyBytes = this.credential.getKey().getBytes(StandardCharsets.UTF_8);
-            byte[] masterKeyDecodedBytes = Utils.Base64Decoder.decode(masterKeyBytes);
-            SecretKey signingKey = new SecretKeySpec(masterKeyDecodedBytes, "HMACSHA256");
-            try {
-                Mac macInstance = Mac.getInstance("HMACSHA256");
-                macInstance.init(signingKey);
-                //  Update the master key hash code
-                this.masterKeyHashCode = masterKeyLatestHashCode;
-                return macInstance;
-            } catch (NoSuchAlgorithmException | InvalidKeyException e) {
-                throw new IllegalStateException(e);
-            }
-        } else {
-            //  Master key hasn't changed, return the cloned mac instance
-            try {
-                return (Mac)this.macInstance.clone();
-            } catch (CloneNotSupportedException e) {
-                throw new IllegalStateException(e);
+            synchronized (this.credential) {
```

Comment:
I am concerned about this synchronized code piece, since this will be called on every request. My concern is this will slow down things. 
Do we need `synchronized` block here even though `credential` is volatile now ?

Reply:
this is very valid point. I also think will have perf impact

Reply:
Only when the hashcode is different is when this code path hits. 
Few triggering points are 
- new client creation starting point
- key-changed

If the assumption is that its a both credentials are valid for a good period of time, the refresh can be asynchronous. 
Can we make that assumption?

Reply:
Yes, that is correct assumption. We can proceed with this PR. looks good to me.

Reply:
Refreshed the implementation to eventually move to the new key. 
One thread will try to refresh where as others will move forward with current key.